go/vt/vtgate/executor_dml_test.go

```diff
@@ -388,6 +388,202 @@ func TestInsertSharded(t *testing.T) {
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(2),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(3),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(4),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(5),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(6),
+		},
+	}, {
+		Sql: "insert ignore into ins_lookup(fromcol, tocol) values (:fromcol0, :tocol0), (:fromcol1, :tocol1), (:fromcol2, :tocol2), (:fromcol3, :tocol3), (:fromcol4, :tocol4)",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol0": sqltypes.Int64BindVariable(1),
+			"tocol0":   sqltypes.Uint64BindVariable(1),
+			"fromcol1": sqltypes.Int64BindVariable(3),
+			"tocol1":   sqltypes.Uint64BindVariable(1),
+			"fromcol2": sqltypes.Int64BindVariable(4),
+			"tocol2":   sqltypes.Uint64BindVariable(1),
+			"fromcol3": sqltypes.Int64BindVariable(5),
+			"tocol3":   sqltypes.Uint64BindVariable(1),
+			"fromcol4": sqltypes.Int64BindVariable(6),
+			"tocol4":   sqltypes.Uint64BindVariable(3),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(1),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(3),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(4),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(5),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(6),
+			"tocol":   sqltypes.Uint64BindVariable(3),
+		},
+	}}
+	if !reflect.DeepEqual(sbclookup.Queries, wantQueries) {
+		t.Errorf("sbclookup.Queries: \n%+v, want \n%+v", sbclookup.Queries, wantQueries)
+	}
+
+	// Test the 0 rows case,
```

Comment:
Wanted to resolve this before addressing the other issues.

The problem here is that this use case breaks a cardinal rule: a row must have one and only one kesypace id. This will make the data essentially corrupt. Here's a (bad) example: Let's assume you have a table with two unique and identical vindexes. This essentially requires the values to be the same. So, the only valid inserts are `insert(1, 1)`, `insert(2, 2)`, etc. If you tried to `insert(1, 2)`, it would fail. But if we decided to allow this to be inserted in the case of an `ignore`, it would go through, probably to shard of `col1`. But if you later ran a query with `col2=2`, you will get no result eventhough such a row exists in the database, and all kinds of other bad things start to happen once data becomes corrupt like this.

If you imagined the sharding rule to be as sacrosanct as a relational rule, such an insert (with two inconsistent unique keys) would not succeed with mysql. So, we should not allow it either.

The case of upsert is more intricate, but still worth disallowing: MySQL says that it will update "one of the rows without breaking any relational rule". In such situations, if we allowed an upsert to pass through, it may succeed at inserting a row thereby corrupting the data. The correct way for us to implement this would be to send an update-only statement instead of the upsert to the correct shard. This is a behavior we can add in the future. So, if you had something like this: `insert into t values(1, 2) on duplicate key update col3=4`, then vitess will find the inconsistency, and should instead send: `update t set col3=4 where col1=1`. For now, the only safe thing we can do is to disallow the forwarding.